File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.py
from collections import Counter

def solution(topping):
    split_count = 0
    topping_counter = Counter(topping)
    half_topping_set = set()
    
    for v in topping:
        half_topping_set.add(v)
        topping_counter[v] -= 1
        
        if topping_counter[v] == 0: topping_counter.pop(v)
        if len(half_topping_set) == len(topping_counter): split_count += 1
        
    return split_count

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.js
function solution(topping) {
    let answer = 0;
    const brother = {}
    const chulsu = new Set()
    
    let broKind = 0

    for (let v of topping) {
        if (brother[v]) brother[v]++
        else {
            brother[v] = 1;
            broKind++
        }
    }
    
    for (let v of topping) {
        chulsu.add(v)
        brother[v]--

        if (!brother[v]) broKind--
        if (chulsu.size === broKind) answer++
    }

    return answer;
}